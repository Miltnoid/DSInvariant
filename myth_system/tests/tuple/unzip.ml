type bool =
  | T
  | F

type list =
  | Nil
  | Cons of bool * list

type pairlist =
  | PNil
  | PCons of (bool * bool) * pairlist

let unzip : pairlist -> list * list |>
{ PNil                 => (Nil, Nil)
| PCons ((T, T), PNil) => ([T], [T])
| PCons ((T, F), PNil) => ([T], [F])
| PCons ((F, T), PNil) => ([F], [T])
| PCons ((F, F), PNil) => ([F], [F])
| PCons ((F, T), PCons ((T, T), PNil)) => ([F; T], [T; T])
| PCons ((F, F), PCons ((F, T), PNil)) => ([F; F], [F; T])
} = ?
