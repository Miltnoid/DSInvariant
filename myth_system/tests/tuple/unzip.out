(* Generated Wed Dec 31 14:27:29 2014 *)
let unzip : pairlist -> list * list =
  let rec f1 (p1:pairlist) : list * list =
    match p1 with
      | PNil -> (Nil, Nil)
      | PCons ((b1, b2), p2) -> (Cons (b1, #1 (f1 p2)),
                                  Cons (b2, #2 (f1 p2)))
  in
    f1
;;
